native/src/widget/grid.rs

```diff
@@ -0,0 +1,276 @@
+
+                    if let Some(column_width) = column_widths.get_mut(column) {
+                        *column_width = column_width.max(layout.width);
+                    } else {
+                        column_widths.insert(column, layout.width);
+                    }
+                }
+
+                // list of x alignments for every column
+                let column_aligns = iter::once(&0.)
+                    .chain(column_widths.iter().take(column_widths.len() - 1))
+                    .scan(0., |state, width| {
+                        *state += width;
+                        Some(*state)
+                    });
+
+                let mut nodes = Vec::with_capacity(self.elements.len());
+                let mut grid_height = 0.;
+                let mut row_height = 0.;
+
+                for ((column, column_align), size) in
+                    column_aligns.enumerate().cycle().zip(layouts)
+                {
+                    if column == 0 {
+                        grid_height += row_height;
+                        row_height = 0.;
+                    }
+
+                    let mut node = Node::new(size);
+                    node.move_to(Point::new(column_align, grid_height));
+                    nodes.push(node);
+                    row_height = row_height.max(size.height);
+                }
+
+                grid_height += row_height;
+                let grid_width = column_widths.into_iter().sum();
+
+                Node::with_children(Size::new(grid_width, grid_height), nodes)
+            // if we have `column_width` but no `columns`, calculate number of
+            // columns by checking how many can fit
+            } else if let Some(column_width) = self.column_width {
+                let column_width: f32 = column_width.into();
+                let max_width = limits.max().width;
+                let columns = (max_width / column_width).floor() as usize;
+                let mut nodes = Vec::with_capacity(self.elements.len());
+                let mut grid_height = 0.;
+                let mut row_height = 0.;
+
+                for (column, element) in
+                    (0..columns).cycle().zip(&self.elements)
+                {
+                    if column == 0 {
+                        grid_height += row_height;
+                        row_height = 0.;
+                    }
+
+                    let size = element.layout(renderer, &column_limits).size();
+                    let mut node = Node::new(size);
+                    node.move_to(Point::new(
+                        column as f32 * column_width,
+                        grid_height,
+                    ));
+                    nodes.push(node);
+                    row_height = row_height.max(size.height);
+                }
+
+                grid_height += row_height;
+                let grid_width = (columns as f32) * column_width;
+
+                Node::with_children(Size::new(grid_width, grid_height), nodes)
```

Comment:
Maybe we should split this in a couple of private functions.

Reply:
Done.